File: Machine.cpp
#include "Machine.h"
#include <string>
#include <iostream>

using namespace std;

Rotor Machine::r1;
Rotor Machine::r2;
Rotor Machine::r3;
Rotor Machine::reflector;

string secretR1 = "dgnotpubfkacivhlsmrjwyxzeq";
string secretR2 = "lompwdzfsravkejnixtgchyqbu";
string secretR3 = "qkgslyfbphotwuraemnczdijvx";
string secretReflector = "lrufxmntzevkphiqodaywbjgsc";

Machine::Machine() {
    this->rotorOneDegree = 0;
    this->rotorTwoDegree = 0;
    this->rotorThreeDegree = 0;
}

Machine::Machine(bool secretModeFlag) {
    r1.setSecretAlphabet(secretR1);
    r2.setSecretAlphabet(secretR2);
    r3.setSecretAlphabet(secretR3);
    reflector.setSecretAlphabet(secretReflector);
    this->rotorOneDegree = 0;
    this->rotorTwoDegree = 0;
    this->rotorThreeDegree = 0;
}

void Machine::resetRotorDegrees() {
    rotorOneDegree = 0;
    rotorTwoDegree = 0;
    rotorThreeDegree = 0;
}

/* Increments the first rotor by one. The second rotor is incremented when 
 * the first makes a full rotation and the third is incremented when the second
 * makes a full rotation. */
void Machine::incrementRotorDegrees() {
    if(rotorOneDegree == 26) {
        if(rotorTwoDegree == 26) {
            if(rotorThreeDegree == 26) {
                /* All three rotors have made a full rotation, so
                 * reset them completely. */
                resetRotorDegrees();
            } else {
                rotorThreeDegree++;
                rotorTwoDegree = 1;
            }
        } else {
            rotorTwoDegree++;
            rotorOneDegree = 1;
        } 
    } else {
        rotorOneDegree++;
    }
}

char Machine::transformCharacter(char c, bool encode) {
    char r1c = r1.getEncodedChar(c, rotorOneDegree);
    char r2c = r2.getEncodedChar(r1c, rotorTwoDegree);
    char r3c = r3.getEncodedChar(r2c, rotorThreeDegree);
    char reflectedr3c;
    if(encode) {
        /* Pass in 0 for rotation degree because the reflector doesn't rotate. */
        reflectedr3c = this->reflector.getEncodedChar(r3c, 0); 
    } else {
        reflectedr3c = this->reflector.getDecodedChar(r3c, 0); 
    }
    char r3dc = r3.getDecodedChar(reflectedr3c, rotorThreeDegree);
    char r2dc = r2.getDecodedChar(r3dc, rotorTwoDegree);
    char r1dc = r1.getDecodedChar(r2dc, rotorOneDegree);
    return r1dc; 
}

std::string Machine::getTransformedString(std::string word, bool encode) {
    std::string encodedWord;
    
    if(word == "0") {
        /* Reset rotors when given end of sentence signal ("0"). We are 
         * done encoding the sentence. */
        resetRotorDegrees();
        return "0";
    } else {
    
        for(int i = 0; i < word.length(); i++) {
            char c = word.at(i);
            encodedWord.append(1, transformCharacter(c, encode));
            incrementRotorDegrees();
        }
        
        return encodedWord;
    }
};

File: pythonVersion/machine.py
from rotor import Rotor
import charChecker

class Machine():

    alphabet = "abcdefghijklmnopqrstuvwxyz";
    digits = "0123456789";

    #not very secure, I know...
    secretR1 = "dgnotpubfkacivhlsmrjwyxzeq";
    secretR2 = "lompwdzfsravkejnixtgchyqbu";
    secretR3 = "qkgslyfbphotwuraemnczdijvx";
    secretReflector = "lrufxmntzevkphiqodaywbjgsc";

    secretD1 = "3852069147"
    secretD2 = "0678124359"
    secretD3 = "7348910625"
    secretDReflector = "4689517320"

    #r1 = Rotor();
    #r2 = Rotor();
    #r3 = Rotor();
    #reflector = Rotor();

    r1degree = 0;
    r2degree = 0;
    r3degree = 0;
    
    def resetRotorDegrees(self):
        self.r1degree = 0;
        self.r2degree = 0;
        self.r3degree = 0;


    def incrementRotorDegrees(self):
        if(self.r1degree == self.alphLength):
            if(self.r2degree == self.alphLength):
                if(self.r3degree == self.alphLength):
                    self.resetRotorDegrees();
                else:
                    self.r3degree = self.r3degree + 1;
                    self.r2degree = 1;
            else:
                self.r2degree = self.r2degree + 1;
                self.r1degree = 1;
        else:
            self.r1degree = self.r1degree + 1;			


    def transformCharacter(self, c, encode):

        if (self.digitEncoding and not c in self.digits) or (not self.digitEncoding and not c in self.alphabet):
            if charChecker.checkCharAcceptable(c):
                return c;
            else:
                return "#";
 
        r1c = self.r1.getEncodedChar(c, self.r1degree);
        r2c = self.r2.getEncodedChar(r1c, self.r2degree);
        r3c = self.r3.getEncodedChar(r2c, self.r3degree);

        if(encode):
            reflectorC = self.reflector.getEncodedChar(r3c, 0);
        else:
            reflectorC = self.reflector.getDecodedChar(r3c, 0);

        r3dc = self.r3.getDecodedChar(reflectorC, self.r3degree);
        r2dc = self.r2.getDecodedChar(r3dc, self.r2degree);
        r1dc = self.r1.getDecodedChar(r2dc, self.r1degree);
        return r1dc;


    def getTransformedString(self, word, encode):
        if word == "<>":
            self.resetRotorDegrees();
            return "<>";

        newWord = "";
        for char in word:	
            newWord = newWord + self.transformCharacter(char, encode);	
            self.incrementRotorDegrees();
    
        return newWord;
        
    def __init__(self, letterFlag, secretModeFlag):
        self.digitEncoding = not letterFlag;
        if letterFlag:
            self.alphLength = len(self.alphabet);
        else:
            self.alphLength = len(self.digits);

        if(letterFlag):
            self.r1 = Rotor(self.alphabet);
            self.r2 = Rotor(self.alphabet);
            self.r3 = Rotor(self.alphabet);
            self.reflector = Rotor(self.alphabet);
        else:
            self.r1 = Rotor(self.digits);
            self.r2 = Rotor(self.digits);
            self.r3 = Rotor(self.digits);
            self.reflector = Rotor(self.digits);

        if(secretModeFlag):
            if(letterFlag):
                self.r1.setSecretAlphabet(self.secretR1);	
                self.r2.setSecretAlphabet(self.secretR2);	
                self.r3.setSecretAlphabet(self.secretR3);	
                self.reflector.setSecretAlphabet(self.secretReflector);	
            else:
                self.r1.setSecretAlphabet(self.secretD1);	
                self.r2.setSecretAlphabet(self.secretD2);	
                self.r3.setSecretAlphabet(self.secretD3);	
                self.reflector.setSecretAlphabet(self.secretDReflector);	

                
